Sources/ApolloCodegenLib/Frontend/GraphQLSchema.swift

```diff
@@ -0,0 +1,119 @@
+import JavaScriptCore
+
+/// A GraphQL schema.
+public class GraphQLSchema: JavaScriptObject {
+  private var typeMap: [String: GraphQLNamedType] = [:]
+  
+  func getType(named typeName: String) throws -> GraphQLNamedType? {
+    if let type = typeMap[typeName] {
+      return type
+    }
+    
+    let type: GraphQLNamedType = try invokeMethod("getType", with: typeName)
+    typeMap[typeName] = type
+    return type
+  }
+  
+  func getPossibleTypes(_ abstractType: GraphQLAbstractType) throws -> [GraphQLObjectType] {
+    return try invokeMethod("getPossibleTypes", with: abstractType)
+  }
+  
+  func getImplementations(interfaceType: GraphQLInterfaceType) throws -> InterfaceImplementations {
+    return try invokeMethod("getImplementations", with: interfaceType)
+  }
+  
+  class InterfaceImplementations: JavaScriptObject {
+    private(set) lazy var objects: [GraphQLObjectType] = self["objects"]
+    private(set) lazy var interfaces: [GraphQLInterfaceType] = self["interfaces"]
+  }
+    
+  func isSubType(abstractType: GraphQLAbstractType, maybeSubType: GraphQLNamedType) throws -> Bool {
+    return try invokeMethod("isSubType", with: abstractType, maybeSubType)
+  }
+}
+
+public class GraphQLNamedType: JavaScriptObject {
+  private(set) lazy var name: String = self["name"]
```

Comment:
Does this just crash if there's no `"name"` field on the JSON object here?

Reply:
No. JavaScript is very loosely typed, and its type conversion rules can be a little surprising... If `name` isn't defined on the JavaScript object, it would actually return the literal string `"undefined"` here. (If you made the property optional in Swift, you would get `nil` in that case.)

Reply:
Is there a way to validate that `self["name"]` isn't returning the `"undefined"` string literal? I feel like that could cause all sorts of weirdness if not properly handled.

Reply:
Yep, I already fixed this in https://github.com/apollographql/apollo-ios/commit/5967af180a9a434107e858372b2419ff9600d24b.